### DISCUSSION TITLE
Suppress output

### DISCUSSION THREAD

**Discussion by sheldonhull** [magefile/mage#372] [Q&A] [answered]
I'm probably doing something wrong, but if I have mage run:

```go
	if err = sh.Run("go", "mod", "tidy"); err != nil {
		return err
	}
```

I get stdout. I've tried other commands in the mage sh package but all output to stdout even without the `-v` flag it seems. 

Do you have any example of writing a custom multiwriter with Mage so I can redirect stdout and only output on error?
Trying to trim the initialization output for local work when no error is being detected if I can.

**Comment by natefinch** [comment]
Are you sure you don't have MAGEFILE_VERBOSE set?  sh.Run shouldn't output to stdout without that or -v.

**Reply by sheldonhull** [reply]
@natefinch yes, I thought that would be the expected behavior, but it still outputs. 

Let's say I have a command that loops through a list of go modules to install as cli tools. 

```go

	// as of last time I checked 2021-09, go get/install wasn't noted to be safe to run in parallel, so keeping it simple with just loop
	for i, t := range tools {
		msg := fmt.Sprintf("install [%d] %s]", i, t)
		// spinner, _ := pterm.DefaultSpinner.
		// 	WithSequence("|", "/", "-", "|", "/", "-", "\\").
		// 	WithRemoveWhenDone(true).
		// 	WithText(msg).
		// 	WithShowTimer(true).Start()

		err := sh.RunWith(env, "go", append(args, t)...)
		if err != nil {
			pterm.Warning.Printf("Could not install [%s] per [%v]\n", t, err)
			// spinner.Fail(fmt.Sprintf("Could not install [%s] per [%v]\n", t, err))

			continue
		}
		// spinner.Success(msg)
		pterm.Success.Println(msg)
	}
```

I stripped out all the progressbar/spinner stuff as when running through a list using go install it streamed all the github packages the go tool was installing. I tried with Run as well. No luck. I'm not setting any environment variable here or passing `-v`. Any ideas?

![2021-09-25-16 11 44-output](https://user-images.githubusercontent.com/3526320/134786045-9b16edef-8073-4b7b-b516-83e287d64b3b.png)

**Reply by natefinch** [reply]
Are you sure that's stdout and not stderr?  Lots of things log to stderr.

**Reply by sheldonhull** [reply]
Got ya. That's probably what's snagging me. I'm betting go outputs to stderr for this. 
Not a big deal, but was hoping to have a simple way to suppress the normal output unless an error occurred. 

Without capturing and parsing string of the output, do you know of any simple way to suppress the output from go install/go mod tidy commands unless an error occurs? I couldn't find any "quiet" flag and guessing there isn't option to do this.

**Comment by natefinch** [comment, marked as answer]
You can use os/exec.Command :)
